### ISSUE TITLE
mDNS queries not working with known answers appended.

### ISSUE THREAD

**Issue by vchandap** [systemd/systemd#29168]
### systemd version the issue has been seen with

252

### Used distribution

yocto open embedded 

### Linux kernel version used

5.4.234

### CPU architectures issue was seen on

arm

### Component

systemd-resolved

### Expected behaviour you didn't see

mDNS quires with known answer appended should be processed without any error.

### Unexpected behaviour you saw
The issue is systemd mdns publisher/resolver rejecting these queries appended with known answers with error "Received mDNS query without question, ignoring".

### Steps to reproduce the problem

Use two devices one as systemd querier (Device A) and other one as systemd mdns publisher/resolver (Device B).

1.Enable systemd mDNS on both devices

2.Add dnssd service with name "_test._udp.local" in Device B

3.Restart systemd services on both devices to make sure caches are flushed and dnssd service is populated.

4.Issue the 1st query from Device A
resolvectl query --protocol=mdns --interface=eth0 --type=PTR _test._udp.local

5.Issue the 2nd query from Device A **before TTL expired in 2mintues with cache disabled flag.**
resolvectl query --protocol=mdns --interface=eth0 --cache=false --type=PTR _test._udp.local

### Additional program output to the terminal or log subsystem illustrating the issue

Observation from tcpdump 
1.Truncked flag is enabled even when a known answer is appended successfully in 1st packet (Packet No#3 in tcpdump).

![image](https://github.com/systemd/systemd/assets/99871235/6f5aafb7-486d-40e8-a776-e3f7419fce70)

2.A subsequent empty query without question and answer sent from systemd querier. (Packet No#4 in tcpdump).
![image](https://github.com/systemd/systemd/assets/99871235/50da8563-8d89-4d6a-bf3c-265451da2dbb)

**Comment by vchandap** [comment]
we further analyzed and found the root cause.  
 
This behavior caused by this PR #23875.
 
The commit https://github.com/systemd/systemd/pull/23875/commits/325513bc776c739a814996cc5c483235ca92be86 enforcing packet truncations (TC bit) and at same time systemd mDNS responder doesn't support mDNS packet truncation https://github.com/systemd/systemd/blob/5898cef22a35ceefa068d5f46929eced2baab0ed/src/resolve/resolved-dns-packet.c#L310 Hence queries with known answers are not working as expected.
  

https://github.com/systemd/systemd/blob/5898cef22a35ceefa068d5f46929eced2baab0ed/src/resolve/resolved-dns-cache.c#L1289

 @yuwata the effect of this logic you have added in this function seems to be that a new packet is created and added to the chain even though the existing packet has the capacity to hold more answers.

This causes the TC flag being set in the original packet, followed by an empty packet. Is there a reason to do this?

**Comment by vieira** [comment]
Hello, I believe I'm experiencing the same issue when doing a reverse mDNS lookup with both the client and server using `systemd-resolved` and `resolvectl` **252.17** on debian 12 (bookworm).

The queries below are made from alpha.local (10.0.0.7) to pi.local (10.0.0.5). The lookup for the name works but the reverse lookup for the IP fails.

```
$ resolvectl query pi.local
pi.local: 10.0.0.5                             -- link: eth0

-- Information acquired via protocol mDNS/IPv4 in 2.9ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: network
```

```
$ resolvectl query 10.0.0.5
10.0.0.5: resolve call failed: All attempts to contact name servers or networks failed
```

I changed the log-level to debug in the "client" (alpha.local, 10.0.0.7) and observed the following (note the `Sending truncated (!) query packet`).

```
alpha systemd-resolved[44551]: D-Bus address resolution request from client PID 45235 (n/a) with UID 4294967295
alpha systemd-resolved[44551]: Looking up RR for 5.0.0.10.in-addr.arpa IN PTR.
alpha systemd-resolved[44551]: Cache miss for 5.0.0.10.in-addr.arpa IN PTR
alpha systemd-resolved[44551]: Firing regular transaction 31893 for <5.0.0.10.in-addr.arpa IN PTR> scope mdns on eth0/INET (validate=yes).
alpha systemd-resolved[44551]: Delaying mdns transaction 31893 for 0us.
alpha systemd-resolved[44551]: Initial jitter phase for transaction 31893 elapsed.
alpha systemd-resolved[44551]: Retrying transaction 31893.
alpha systemd-resolved[44551]: Cache miss for 5.0.0.10.in-addr.arpa IN PTR
alpha systemd-resolved[44551]: Firing regular transaction 31893 for <5.0.0.10.in-addr.arpa IN PTR> scope mdns on eth0/INET (validate=yes).
alpha systemd-resolved[44551]: Sending truncated (!) query packet with id 0 on interface 2/AF_INET of size 827.
alpha systemd-resolved[44551]: Sending query packet with id 0 on interface 2/AF_INET of size 12.
alpha systemd-resolved[44551]: Timeout reached on transaction 31893.

alpha systemd-resolved[44551]: Retrying transaction 31893.
alpha systemd-resolved[44551]: Cache miss for 5.0.0.10.in-addr.arpa IN PTR
alpha systemd-resolved[44551]: Firing regular transaction 31893 for <5.0.0.10.in-addr.arpa IN PTR> scope mdns on eth0/INET (validate=yes).
alpha systemd-resolved[44551]: Sending truncated (!) query packet with id 0 on interface 2/AF_INET of size 827.
alpha systemd-resolved[44551]: Sending query packet with id 0 on interface 2/AF_INET of size 12.
alpha systemd-resolved[44551]: Timeout reached on transaction 31893.
```

I also changed the log-level to debug in the destination and observed the following errors (note the `Received mDNS query without question, ignoring.`):
```
pi systemd-resolved[2882]: Received mdns UDP packet of size 145, ifindex=2, ttl=255, fragsize=0, sender=10.0.0.7, destination=224.0.0.251
pi systemd-resolved[2882]: Invalid mDNS UDP packet.
pi systemd-resolved[2882]: Received mdns UDP packet of size 12, ifindex=2, ttl=255, fragsize=0, sender=10.0.0.7, destination=224.0.0.251
pi systemd-resolved[2882]: Got mDNS query packet for id 0
pi systemd-resolved[2882]: Received mDNS query without question, ignoring.
pi systemd-resolved[2882]: mDNS query processing failed: Bad message
pi systemd-resolved[2882]: Received mdns UDP packet of size 145, ifindex=2, ttl=255, fragsize=0, sender=10.0.0.7, destination=224.0.0.251
pi systemd-resolved[2882]: Invalid mDNS UDP packet.
pi systemd-resolved[2882]: Received mdns UDP packet of size 12, ifindex=2, ttl=255, fragsize=0, sender=10.0.0.7, destination=224.0.0.251
pi systemd-resolved[2882]: Got mDNS query packet for id 0
pi systemd-resolved[2882]: Received mDNS query without question, ignoring.
pi systemd-resolved[2882]: mDNS query processing failed: Bad message
pi systemd-resolved[2882]: Received mdns UDP packet of size 145, ifindex=2, ttl=255, fragsize=0, sender=10.0.0.7, destination=224.0.0.251
pi systemd-resolved[2882]: Invalid mDNS UDP packet.
pi systemd-resolved[2882]: Received mdns UDP packet of size 12, ifindex=2, ttl=255, fragsize=0, sender=10.0.0.7, destination=224.0.0.251
pi systemd-resolved[2882]: Got mDNS query packet for id 0
pi systemd-resolved[2882]: Received mDNS query without question, ignoring.
```

I also noticed that this happens only when the query is sent by systemd-resolved to a destination that also runs systemd-resolved. As a control I also have another box running debian 12 but with `avahi-daemon` instead: lambda.local (10.0.0.6):

```
resolvectl query 10.0.0.6 (running avahi-daemon)
10.0.0.6: lambda.local                         -- link: eth0

-- Information acquired via protocol mDNS/IPv4 in 53.6ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: network
```

and the log shows:

```
alpha systemd-resolved[44551]: Cache miss for 6.0.0.10.in-addr.arpa IN PTR
alpha systemd-resolved[44551]: Firing regular transaction 31941 for <6.0.0.10.in-addr.arpa IN PTR> scope mdns on eth0/INET (validate=yes).
alpha systemd-resolved[44551]: Sending truncated (!) query packet with id 0 on interface 2/AF_INET of size 123.
alpha systemd-resolved[44551]: Sending query packet with id 0 on interface 2/AF_INET of size 12.
alpha systemd-resolved[44551]: Received mdns UDP packet of size 59, ifindex=2, ttl=255, fragsize=0, sender=10.0.0.6, destination=224.0.0.251
alpha systemd-resolved[44551]: Got mDNS reply packet
alpha systemd-resolved[44551]: Checking for conflicts...
alpha systemd-resolved[44551]: Processing incoming packet of size 59 on transaction 31941 (rcode=SUCCESS).
alpha systemd-resolved[44551]: Regular transaction 31941 for <6.0.0.10.in-addr.arpa IN PTR> on scope mdns on eth0/INET now complete with <success> from network (unsigned; non-confidential).
alpha systemd-resolved[44551]: Freeing transaction 31941.
alpha systemd-resolved[44551]: Added positive unauthenticated non-confidential cache entry for 6.0.0.10.in-addr.arpa IN PTR 120s on eth0/INET/10.0.0.6
```

Also using dig instead of systemd-resolved to fire the query seems to work:

```
dig -p 5353 @224.0.0.251 -x 10.0.0.5

; <<>> DiG 9.10.6 <<>> -p 5353 @224.0.0.251 -x 10.0.0.5
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 32180
;; flags: qr aa; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;5.0.0.10.in-addr.arpa.		IN	PTR

;; ANSWER SECTION:
5.0.0.10.in-addr.arpa.	120	IN	PTR	pi.local.

;; Query time: 12 msec
;; SERVER: 10.0.0.5#5353(224.0.0.251)
;; WHEN: Thu Oct 19 20:20:48 WEST 2023
;; MSG SIZE  rcvd: 61
```

and the systemd-resolved debug log in this case shows:

```
pi systemd-resolved[2882]: Received mdns UDP packet of size 50, ifindex=2, ttl=1, fragsize=0, sender=10.0.0.128, destination=224.0.0.251
pi systemd-resolved[2882]: Got mDNS query packet for id 27000
pi systemd-resolved[2882]: Sending response packet with id 27000 on interface 2/AF_INET of size 61.
```

@yuwata Do you think this problem has the same root cause of the issue reported here or should I open another ticket?

**Comment by yuwata** [comment]
Thank you for the report, and sorry for late response. Fix is waiting in #29927. If possible, please test the PR.
